tensorflow_recommenders_addons/dynamic_embedding/python/ops/warm_start_util.py

```diff
@@ -0,0 +1,109 @@
+
+# lint-as: python3
+"""warm-start util"""
+
+import collections
+
+import six
+
+from tensorflow.python.framework import errors
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.ops import variables as variables_lib
+from tensorflow.python.platform import tf_logging as logging
+from tensorflow.python.training import checkpoint_ops
+from tensorflow.python.training import checkpoint_utils
+from tensorflow.python.training import saver as saver_lib
+from tensorflow.python.training.saving import saveable_object_util
+from tensorflow.python.util.tf_export import tf_export
+from tensorflow.python.training import checkpoint_utils
+from tensorflow.python.training import warm_starting_util
+from tensorflow.python.training import saver as saver_lib
+from tensorflow.python.ops import io_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import variable_scope
+
+from tensorflow_recommenders_addons import dynamic_embedding as de
+
+
+def _get_de_variables(vars_to_warm_start):
+  if isinstance(vars_to_warm_start,
+                six.string_types) or vars_to_warm_start is None:
+    list_of_vars = op.get_collection(de.GraphKeys.DYNAMIC_EMBEDDING_VARIABLES,
+                                     scope=vars_to_warm_start)
+  elif isinstance(vars_to_warm_start, list):
+    if all(isinstance(v, six.string_types) for v in vars_to_warm_start):
+      list_of_vars = []
+      for v in vars_to_warm_start:
+        list_of_vars += ops.get_collection(
+            de.GraphKeys.DYNAMIC_EMBEDDING_VARIABLES, scope=v)
+    elif all(isinstance(v, de.Variable) for v in vars_to_warm_start):
+      list_of_vars = vars_to_warm_start
+    else:
+      raise ValueError("If `vars_to_warm_start` is a list, it must be a "
+                       "`de.Variable` or `str`.  Given types are {}".format(
+                           type(vars_to_warm_start)))
+
+  de_variables = []
+  for v in list_of_vars:
+    t = [v] if not isinstance(v, list) else v
+    de_variables.append(v)
+
+  return de_variables
+
+
+def warm_start(ckpt_to_initialize_from,
+               vars_to_warm_start=".*",
+               var_name_to_prev_var_name=None):
+  logging.info("Warm-starting from: {}".format(ckpt_to_initialize_from))
+
+  de_variables = _get_de_variables(vars_to_warm_start)
+  if not var_name_to_prev_var_name:
+    var_name_to_prev_var_name = {}
+
+  ckpt_file = checkpoint_utils._get_checkpoint_filename(ckpt_to_initialize_from)
+  assign_ops = []
+  for variable in de_variables:
+    var_name = variable.name
+    prev_var_name = var_name_to_prev_var_name.get(var_name)
+    if prev_var_name:
+      logging.debug('Warm-starting variable: {}; prev_var_name: {}'.format(
+          var_name, prev_var_name or "Unchanged"))
+    else:
+      prev_var_name = var_name
+
+    store_vars = variable_scope._get_default_variable_store()._vars
```

Comment:
It seems that "store_vars" here only stores variables with current var names. So store_vars.get(prev_var_name, None) will always return None when prev_var_name and var name are different?